### PR TITLE
FT connector: improvements, add more tests

### PR DESCRIPTION
**Pull request by karim-en**
- Add a function to remove an account from the whitelist;
- Use `sender_id` instead of `signer_id` in the `ft_on_transfer` to check if the account whitelisted;
- Add FT locker tests to CI;
- Add tests for whitelisting functionality.

### CODE REVIEW

File: erc20-bridge-token/test/BridgeToken.js [hunk 1]
```diff
@@ -27,6 +27,26 @@ describe('BridgeToken', () => {
     await ProofConsumer.transferOwnership(BridgeTokenFactory.address);
   })
 
+  async function setMetadata(nearTokenId, tokenProxyAddress) {
+    const metadataProof = require('./proof_template.json');
+    const metadata = createDefaultERC20Metadata(nearTokenId, 1089);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:32]: Please use constants for the block proof height

File: erc20-bridge-token/test/BridgeToken.js [hunk 2]
```diff
@@ -27,6 +27,26 @@ describe('BridgeToken', () => {
...
+    await expect(
+      BridgeTokenFactory.setMetadata(borshifyOutcomeProof(metadataProof), 1089)
+    )
+      .to.emit(BridgeTokenFactory, "SetMetadata")
+      .withArgs(
+        tokenProxyAddress,
+        metadata.name,
+        metadata.symbol,
+        metadata.decimals
+      );
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:47]:
```suggestion
    ).toString("base64");
    
    await expect(
      BridgeTokenFactory.setMetadata(borshifyOutcomeProof(metadataProof), 1089)
    )
      .to
      .emit(BridgeTokenFactory, "SetMetadata")
      .withArgs(
        tokenProxyAddress,
        metadata.name,
        metadata.symbol,
        metadata.decimals
      );
```

File: erc20-bridge-token/test/BridgeToken.js [hunk 3]
```diff
@@ -46,19 +66,20 @@ describe('BridgeToken', () => {
...
+    const { tokenProxyAddress, token } = await createEmptyToken(
+      nearTokenId,
+      BridgeTokenFactory,
+      BridgeTokenInstance
+    );
+    await setMetadata(nearTokenId, tokenProxyAddress);
+    expect(await token.name()).to.equal("NEAR ERC20");
+    expect(await token.symbol()).to.equal("NEAR");
+    expect((await token.decimals()).toString()).to.equal("18");
+    expect((await token.metadataLastUpdated()).toString()).to.equal("1089");
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:81]:
```suggestion
    );
    
    await setMetadata(nearTokenId, tokenProxyAddress);
    expect(await token.name()).to.equal("NEAR ERC20");
    expect(await token.symbol()).to.equal("NEAR");
    expect((await token.decimals()).toString()).to.equal("18");
    expect((await token.metadataLastUpdated()).toString()).to.equal("1089");
```

File: erc20-bridge-token/test/BridgeToken.js [hunk 4]
```diff
@@ -105,13 +126,15 @@ describe('BridgeToken', () => {
...
-    const { token } = await createEmptyToken(nearTokenId, BridgeTokenFactory, BridgeTokenInstance)
-    const metadataProof = require('./proof_template.json');
-    metadataProof.outcome_proof.outcome.status.SuccessValue = serialize(SCHEMA, 'SetMetadataResult', createDefaultERC20Metadata(nearTokenId, 1089)).toString('base64');
-    await BridgeTokenFactory.setMetadata(borshifyOutcomeProof(metadataProof), 1089);
+    const { tokenProxyAddress, token } = await createEmptyToken(
+      nearTokenId,
+      BridgeTokenFactory,
+      BridgeTokenInstance
+    );
+    setMetadata(nearTokenId, tokenProxyAddress);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:134]:
```suggestion
    );
    
    setMetadata(nearTokenId, tokenProxyAddress);
```

File: erc20-bridge-token/test/BridgeToken.js [hunk 5]
```diff
@@ -159,41 +180,51 @@ describe('BridgeToken', () => {
...
+    const lockResultProof = require('./proof_template.json');
     lockResultProof.outcome_proof.outcome.status.SuccessValue = serialize(SCHEMA, 'LockResult', {
       prefix: RESULT_PREFIX_LOCK,
       token: nearTokenId,
-      amount: 100,
+      amount: amountToTransfer,
       recipient: ethers.utils.arrayify(adminAccount.address),
     }).toString('base64');
     lockResultProof.outcome_proof.outcome.receipt_ids[0] = 'D'.repeat(44);
     await BridgeTokenFactory.setTokenWhitelistMode(nearTokenId, 2);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:203]: Please use constants for whitelisting modes in tests

File: erc20-bridge-token/test/BridgeToken.js [hunk 6]
```diff
@@ -159,41 +180,51 @@ describe('BridgeToken', () => {
...
-    await BridgeTokenFactory.deposit(borshifyOutcomeProof(metadataProof), 1090);
+    await BridgeTokenFactory.deposit(borshifyOutcomeProof(lockResultProof), 1090);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:204]: Please use a constant for block proof height here and in all other cases

File: erc20-bridge-token/test/BridgeToken.js [hunk 7]
```diff
@@ -298,4 +322,118 @@ describe('BridgeToken', () => {
...
+    let testProofId = 1;
+    const recipient = "testrecipient.near";
+
+    beforeEach(async function() {
+      tokenInfo = await createEmptyToken(
+        nearTokenId,
+        BridgeTokenFactory,
+        BridgeTokenInstance
+      );
+      setMetadata(nearTokenId, tokenInfo.tokenProxyAddress);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:337]:
```suggestion
      );
      
      setMetadata(nearTokenId, tokenInfo.tokenProxyAddress);
```

File: erc20-bridge-token/test/BridgeToken.js [hunk 8]
```diff
@@ -298,4 +322,118 @@ describe('BridgeToken', () => {
...
+        SCHEMA,
+        "LockResult",
+        {
+          prefix: RESULT_PREFIX_LOCK,
+          token: nearTokenId,
+          amount: 100,
+          recipient: ethers.utils.arrayify(adminAccount.address),
+        }
+      ).toString("base64");
+      const source = "123456789FGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:350]: Please name the variable accordingly (maybe `base58Alphabet`?)

File: erc20-bridge-token/test/BridgeToken.js [hunk 9]
```diff
@@ -298,4 +322,118 @@ describe('BridgeToken', () => {
...
+        1090
+      );
+    });
+
+    it("Test account in whitelist", async function() {
+      await BridgeTokenFactory.setTokenWhitelistMode(nearTokenId, 3);
+      await BridgeTokenFactory.addAccountToWhitelist(
+        nearTokenId,
+        adminAccount.address
+      );
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:366]: IMO it's better to use a non-admin account to test the behavior and also perform appropriate further operations (deposit, withdraw) from the user account too.

File: erc20-bridge-token/test/BridgeToken.js [hunk 10]
```diff
@@ -298,4 +322,118 @@ describe('BridgeToken', () => {
...
+      ).to.be.equal("0");
+    });
+
+    it("Test token in whitelist", async function() {
+      await BridgeTokenFactory.setTokenWhitelistMode(nearTokenId, 2);
+      await BridgeTokenFactory.withdraw(nearTokenId, 100, recipient);
+      expect(
+        (await tokenInfo.token.balanceOf(adminAccount.address)).toString()
+      ).to.be.equal("0");
+    });
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:379]: Let's also add a test that will use several tokens, whitelist one of them, and verify that one works as whitelisted and the other does not. Then the same with some accounts on these tokens.

(Please add the same to Rust tests)

**Review by sept-en — CHANGES_REQUESTED** [review]

File: erc20-bridge-token/test/BridgeToken.js (hunk 1 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: 699afa2dcb032eb54349c0516950d007e028b2b1

File: erc20-bridge-token/test/BridgeToken.js (hunk 6 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: 699afa2dcb032eb54349c0516950d007e028b2b1

File: erc20-bridge-token/test/BridgeToken.js (hunk 8 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: 699afa2dcb032eb54349c0516950d007e028b2b1

File: erc20-bridge-token/test/BridgeToken.js (hunk 5 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js:216]: 90ab0818690a662fbe18d7aaa3e913c98185da97

File: erc20-bridge-token/test/BridgeToken.js (hunk 9 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: 11eea06736ce1a2c09d4118f00d10b4a75b72383

File: erc20-bridge-token/test/BridgeToken.js (hunk 10 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: 0db3db48a358b519e9ff357dba4729856a3c0cb3
and 
d988be855d875e4031f22cfc3a75f4638e305503

File: erc20-bridge-token/test/BridgeToken.js [hunk 11]
```diff
@@ -399,6 +403,109 @@ describe('BridgeToken', () => {
...
+      const whitelistTokens = [
+        "near-token1.near",
+        "near-token2.near",
+        "near-token3.near",
+        "near-token4.near",
+      ];
+
+      const signers = await ethers.getSigners();
+      const whitelistAccounts = signers.slice(0, 3);
+      const blacklistAccounts = signers.slice(4, 7);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:468]: Please use lengths of appropriate account lists for indices.

File: erc20-bridge-token/test/BridgeToken.js [hunk 12]
```diff
@@ -399,6 +403,109 @@ describe('BridgeToken', () => {
...
+        "near-token6.near",
+        "near-token7.near",
+        "near-token8.near",
+      ];
+
+      const tokensInfo = [];
+      for (token of whitelistTokens) {
+        tokensInfo.push(await createToken(token));
+        await deposit(token, amountToTransfer, user.address);
+        await BridgeTokenFactory.setTokenWhitelistMode(token, 2);
```
Comment (sept-en) [inline erc20-bridge-token/test/BridgeToken.js:425]: Please use constants for whitelist modes (everywhere)

**Review by sept-en — CHANGES_REQUESTED** [review]

File: erc20-bridge-token/test/BridgeToken.js (hunk 12 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: cf35ca224be4e69c77bbbb4b295a8a00eb9cd9e7

File: erc20-bridge-token/test/BridgeToken.js (hunk 11 above)
Comment (karim-en) [reply erc20-bridge-token/test/BridgeToken.js]: cf35ca224be4e69c77bbbb4b295a8a00eb9cd9e7

**Review by sept-en — APPROVED** [review]

**Pull request merged**